File: Data Structures & Algorithms/valid-binary-search-tree/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def dfs(node, left, right):
            if not node:
                return True
            t1, t2 = True, True
            if node.left:
                t1 = left < node.left.val < node.val
            if node.right:
                t2 = node.val < node.right.val < right
            if not (t1 and t2):
                return False

            return t1 and t2 and dfs(node.left, left, node.val) and dfs(node.right, node.val, right)
        return dfs(root, -float('inf'), float('inf'))
        

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-6.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} k
     * @return {number}
     */
    // a trivial solution is to just collect all the values in the
    // increasing order. the smallest value is always the left child then the center value and then right child
    // then recursively collect all the values. 
    foo(node, vals) {
        if (!node) return [];
        const leftChildren = this.foo(node.left, vals);
        vals.push(node.val);
        const rightChildren = this.foo(node.right, vals);
        return vals;
    };

    kthSmallest(root, k) {
        let node = root;
        const stack = [];
        let ans = -1;
        let count = 0
        while (node || stack.length > 0) {
            while (node) {
                stack.push(node);
                node = node.left;
            }; // this will stop only when node becomes null

            node = stack.pop(); // get the first non-null node which will be the left most

            ans = node.val; // this will be the lowest value
            count ++;
            if (count == k) {
                return ans
            };
            // move to the right node; does not matter if it is null, next non null node will be popped out
            node = node.right; 
        }
    }
}
